pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -2534,18 +2537,35 @@ def _is_valid_return_type(match: TypeHintMatch, node: nodes.NodeNG) -> bool:
 
     if (
         match.check_return_type_inheritance
-        and isinstance(match.return_type, str)
+        and isinstance(match.return_type, (str, list))
         and isinstance(node, nodes.Name)
     ):
-        ancestor: nodes.ClassDef
-        for infer_node in node.infer():
-            if isinstance(infer_node, nodes.ClassDef):
-                if infer_node.name == match.return_type:
+        if isinstance(match.return_type, str):
+            valid_types = {match.return_type}
+        else:
+            valid_types = {el for el in match.return_type if isinstance(el, str)}
+
+        try:
+            for infer_node in node.infer():
+                if _check_ancestry(infer_node, valid_types):
                     return True
-                for ancestor in infer_node.ancestors():
-                    if ancestor.name == match.return_type:
-                        return True
+        except NameInferenceError:
+            for class_node in node.root().nodes_of_class(nodes.ClassDef):
+                if class_node.name == node.name:
+                    for infer_node in class_node.infer():
+                        if _check_ancestry(infer_node, valid_types):
+                            return True
```

Comment:
Minor: The first if-clause acts like a guard. I think using it with `continue` could make sense.
```suggestion
                if class_node.name != node.name:
                    continue
                for infer_node in class_node.infer():
                    if _check_ancestry(infer_node, valid_types):
                        return True
```